src/cluster_legacy.c

```diff
@@ -1250,8 +1250,67 @@ void clusterInitLast(void) {
     }
 }
 
+void clusterAutoFailoverOnShutdown(void) {
+    if (!nodeIsPrimary(myself) || !server.auto_failover_on_shutdown) return;
+
+    /* Find the first best replica, that is, the replica with the largest offset. */
+    int legacy_replica = 0;
+    client *best_replica = NULL;
+    listIter replicas_iter;
+    listNode *replicas_list_node;
+    listRewind(server.replicas, &replicas_iter);
+    while ((replicas_list_node = listNext(&replicas_iter)) != NULL) {
+        client *replica = listNodeValue(replicas_list_node);
+        /* This is done only when the replica offset is caught up, to avoid data loss.
+         * And 0x80100 is 8.1.0, we only support this feature in this version. */
+        if (replica->repl_data->replica_version < 0x80100) {
+            legacy_replica = 1;
+            best_replica = NULL;
+            break;
+        }
```

Comment:
Shouldn't we loop over all the replicas and make the decision ?

```suggestion
        if (replica->repl_data->replica_version < 0x80100) {
            legacy_replica = 1;
            continue;
        }
```

best_replica is already NULL. So, we can remove that statement as well.

Reply:
The reason we break in this case is that we have a shared replication buffer and we want to send something new over the replication stream that legacy replicas would not understand. They might panic, depending on a config. That's why we only enable the feature if *all* replicas support it.

Reply:
.. and no, `best_replica` is not always NULL here. We may have already set it to a replica. We don't break when we find a good replica because we also want to check that all replicas support the feature.

Reply:
Gotcha! My bad!